### PR TITLE
Local response normalize.

### PR DESCRIPTION
**Pull request by gongweibao**
Fix https://github.com/PaddlePaddle/Paddle/issues/4425

### CODE REVIEW

File: paddle/operators/lrn_op.cc [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    ctx.Output<Tensor>("mid_out")->Resize(x_dim);
+    ctx.ShareLoD("X", /*->*/ "Out");
+  }
+};
+
+class LRNOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  LRNOpMaker(framework::OpProto *proto, framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "The first input of lrn op");
```
Comment (lcy-seso) [inline paddle/operators/lrn_op.cc]: Personally, I think only "the first/second input of X" is not a good comment.
https://github.com/PaddlePaddle/Paddle/issues/4314

`fc op` has a good comment style. https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/fc_op.cc#L123

Comment (gongweibao) [reply paddle/operators/lrn_op.cc]: done

**Comment by lcy-seso** [comment]
The comments still do not follow our comment style in our doc: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/name_convention.md

Please take fc_op as an example.

File: paddle/operators/lrn_op.cc [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("X"), "Input(X) should not be null");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar(framework::GradVarName("mid_out")),
+                            "Input(mid_out@GRAD) should not be null");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar(framework::GradVarName("Out")),
+                            "Input(Out@GRAD) should not be null");
+
+    auto x_dims = ctx.Input<Tensor>("X")->dims();
+    auto *x_g = ctx.Output<framework::Tensor>(framework::GradVarName("X"));
+    x_g->Resize(x_dims);
```
Comment (qingqing01) [inline paddle/operators/lrn_op.cc]: Need to update to the latest code.

File: paddle/operators/lrn_op.cc [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+ (Tensor)Input of lrn op.It must be a 4 rank tenor with NCHW format.
+ )DOC");
+
+    AddOutput("Out", "(Tensor)The output of lrn op");
+    AddOutput("mid_out", R"Doc(
+(Tensor)Middle result of lrn op.It's computed in forward process 
+and also used in backward process.
+    )Doc");
+
+    AddAttr<int>("n", R"DOC(
```
Comment (qingqing01) [inline paddle/operators/lrn_op.cc]: n -> size

File: paddle/operators/lrn_op.h [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+          auto c_out_g = e_out_g.slice(Eigen::array<int, 4>({{m, ch, 0, 0}}),
+                                       Eigen::array<int, 4>({{1, 1, H, W}}));
+
+          i_x_g.device(ctx.GetEigenDevice<Place>()) +=
+              ratio * c_out_g * c_out * i_x / c_mid;
+        }
+      }
+    }
+  }
+};
```
Comment (qingqing01) [inline paddle/operators/lrn_op.h]: GPU的实现， 最好复用kernel: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/function/CrossMapNormalOpGpu.cu

现在这样的循环对于GPU来说效率低。

File: paddle/operators/lrn_op.cc [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        .SetDefault(2.0)
+        .GreaterThan(0.0);
+
+    AddAttr<float>("alpha", R"DOC(
+(float, default 0.0001)alpha is the scale number.
+        )DOC")
+        .SetDefault(0.0001)
+        .GreaterThan(0.0);
+
+    AddAttr<float>("beta", R"DOC(
```
Comment (qingqing01) [inline paddle/operators/lrn_op.cc]: `alpha`, `beta`, `k`这些float类型的参数，觉得最好也用模板来写，可以参考:https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/dropout_op.cc#L40

File: python/paddle/v2/framework/tests/test_lrn_op.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+        self.inputs = {'X': self.x}
+        self.outputs = {'Out': self.out, 'mid_out': self.mid_out}
+        self.attrs = self.get_attrs()
+
+    def test_check_output(self):
+        self.check_output()
+
+    def test_check_grad_normal(self):
+        self.check_grad(['X'], 'Out', max_relative_error=0.12)
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_lrn_op.py]: 如果`max_relative_error `过大，可以尝试double测试~

File: paddle/operators/lrn_op.cc [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    PADDLE_ENFORCE_NOT_NULL(ctx.OutputVar("Out"),
+                            "Output(Out) of LRNOp should not be null.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.OutputVar("mid_out"),
+                            "mid_out(Out) of LRNOp should not be null.");
+
+    auto x_dim = ctx.Input<Tensor>("X")->dims();
+    PADDLE_ENFORCE_EQ(x_dim.size(), 4, "Input(X)'rank of LRNOp should be 4.");
+
+    ctx.Output<Tensor>("Out")->Resize(x_dim);
+    ctx.Output<Tensor>("mid_out")->Resize(x_dim);
```
[the remark this answers is not in the document]
Comment (gongweibao) [reply paddle/operators/lrn_op.cc]: Done.

File: paddle/operators/lrn_op.cc [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+  }
+};
+
+class LRNOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  LRNOpMaker(framework::OpProto *proto, framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", R"DOC(
+ (Tensor)Input of lrn op.It must be a 4 rank tenor with NCHW format.
+ )DOC");
```
[the remark this answers is not in the document]
Comment (gongweibao) [reply paddle/operators/lrn_op.cc]: Done.

File: paddle/operators/lrn_op.cc [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+    AddOutput("Out", "(Tensor)The output of lrn op");
```
[the remark this answers is not in the document]
Comment (gongweibao) [reply paddle/operators/lrn_op.cc]: Done.

File: paddle/operators/lrn_op.cc [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    AddOutput("mid_out", R"Doc(
```
[the remark this answers is not in the document]
Comment (gongweibao) [reply paddle/operators/lrn_op.cc]: Done.

File: paddle/operators/lrn_op.cc (hunk 3 above)
Comment (gongweibao) [reply paddle/operators/lrn_op.cc]: 这是公式里边的变量。

File: paddle/operators/lrn_op.cc (hunk 2 above)
Comment (gongweibao) [reply paddle/operators/lrn_op.cc]: Done.

File: paddle/operators/lrn_op.cc (hunk 5 above)
Comment (gongweibao) [reply paddle/operators/lrn_op.cc]: Done.

File: python/paddle/v2/framework/tests/test_lrn_op.py (hunk 6 above)
Comment (gongweibao) [reply python/paddle/v2/framework/tests/test_lrn_op.py]: Done.

File: paddle/operators/lrn_op.h (hunk 4 above)
Comment (gongweibao) [reply paddle/operators/lrn_op.h]: 我增加了一个ISSUE，并且写到了layer port中，新开PR解决这个问题吧！
https://github.com/PaddlePaddle/Paddle/issues/5066

**Pull request merged**